ProjFS.Mac/PrjFSKext/PrjFSKext/KauthHandler.cpp

```diff
@@ -791,45 +815,51 @@ static bool TrySendRequestAndWaitForResponse(
     }
     Mutex_Release(s_outstandingMessagesMutex);
     
-    // TODO(Mac): Should we pass in the root directly, rather than root->index?
-    //            The index seems more like a private implementation detail.
-    if (!isShuttingDown && 0 != ActiveProvider_SendMessage(root, messageSpec))
+    if (isShuttingDown)
     {
-        // TODO: appropriately handle unresponsive providers
-        
-        *kauthResult = KAUTH_RESULT_DEFER;
-        goto CleanupAndReturn;
+        // Message never inserted into list, don't try to remove it on cleanup.
+        return false;
     }
     
-    while (!message.receivedResponse &&
-           !s_isShuttingDown)
+    errno_t sendError = ActiveProvider_SendMessage(root, messageSpec);
+   
+    Mutex_Acquire(s_outstandingMessagesMutex);
     {
-        Sleep(5, &message);
-    }
+        if (0 != sendError)
+        {
+            // TODO: appropriately handle unresponsive providers
+            *kauthResult = KAUTH_RESULT_DEFER;
+            goto CleanupUnlockAndReturn;
+        }
     
-    if (s_isShuttingDown)
-    {
-        *kauthResult = KAUTH_RESULT_DENY;
-        goto CleanupAndReturn;
-    }
+        while (!message.receivedResponse &&
+           !s_isShuttingDown)
+        {
+            Mutex_Sleep(s_outstandingMessagesMutex, &message, 5 /* seconds */);
```

Comment:
I think the existing call to `msleep` inside our `Sleep` wrapper function already has the ability to sleep a mutex, we're just not passing the mutex pointer through to it. I'd like to avoid having two different sleep implementations if we can avoid it - any way to combine these?

Reply:
I don't feel that "sleep for N seconds" and "atomically drop mutex, await event with timeout, reacquire mutex" are conceptually close enough to warrant sharing an API, and that the former purpose (by passing a `nullptr` for the mutex argument) is particularly obvious, but we can certainly go that way.